[docs] - Add guide for freshness checks (DOC-168) (#20724)
dagster-io/dagster: docs/content/concepts/assets/asset-checks/checking-for-data-freshness.mdx

  ),
  cron_schedule="* * * * *",
)

source_table_freshness_checks = build_freshness_checks_for_non_partitioned_assets(
  assets=[source_tables],
  maximum_lag_minutes=2 * 60,
)

defs = Definitions(
  assets=[source_table_freshness_checks],
  asset_checks=[source_table_freshness_checks],
  schedules=[source_tables_observation_schedule],
)
```

From here, you can [view and manually execute the checks in the Dagster UI](#freshness-checks-in-the-dagster-ui).

---

## Defining freshness checks for materializable assets

In this section, we'll demonstrate how to implement freshness checks for materializable assets. These are assets that are materialized by a Dagster pipeline.

### Step 1: Define the freshness check

In this example, we used the `build_freshness_checks` function to produce an asset check that will fail if an asset’s latest materialization is more than two hours before the current time:

```python
from dagster import build_freshness_checks_for_non_partitioned_assets

@asset
def my_asset():
  ...

asset1_freshness_checks = build_freshness_checks_for_non_partitioned_assets(
  assets=[my_asset], maximum_lag_minutes=2 * 60
)

> ```

Author: I think the OG Notion draft had a TODO for providing a better example for this section. I'll update all the code here when that's finished.

The last step is to include the asset, check, and sensor in a <PyObject object="Definitions" /> object. This enables Dagster tools to load everything we've defined:

```python
defs = Definitions(
  assets=[my_asset],
  asset_checks=[asset1_freshness_checks],
  sensors=[freshness_checks_sensor]
)
```

At this point, the finished code should look like this:

```python
from dagster import (
  asset,
  build_freshness_checks_for_non_partitioned_assets,
  build_sensor_for_freshness_checks,
  Definitions
)

@asset
def my_asset():
  ...

asset1_freshness_checks = build_freshness_checks_for_non_partitioned_assets(
  assets=[my_asset], maximum_lag_minutes=2 * 60
)

freshness_checks_sensor = build_sensor_for_freshness_checks(
  freshness_checks=[asset1_freshness_checks]
)

defs = Definitions(
  assets=[my_asset],
  asset_checks=[asset1_freshness_checks],
  sensors=[freshness_checks_sensor]
)
```

From here, you can [view and manually execute the checks in the Dagster UI](#freshness-checks-in-the-dagster-ui).

---

## Freshness checks in the Dagster UI

To view a freshness check, navigate to the **Asset details** page for the asset and click the **Checks tab**:

> \[TODO - IMAGE]

Author: TODO: Add screenshot(s) to this section

This page allows you to see metadata about the check and its latest execution. You can also manually execute the check by clicking the **Execute** button.

---

## APIs in this guide

> \[TODO - update this list]

Author: TODO: Update list

---
title: "Using asset checks to check data freshness | Dagster Docs"
description: "Use freshness checks to identify data assets that are overdue for an update."
---

# Using asset checks to check data freshness

> A freshness check is a type of [asset check](/concepts/assets/asset-checks) that allows you to define the amount of time between materializations/observations for Dagster [assets](/concepts/assets/software-defined-assets).

Reviewer: "define the amount of time" makes it sound like all freshness checks are based on a fixed time interval between updates, but some of the ways we allow defining freshness checks are a little different. I would lean towards just "A freshness check is a type of asset check that allows you to identify assets that are overdue for an update."

Author: For my own understanding, you're referring to the time anomaly option here, right? Since that doesn't really allow you to define an amount of time for freshness, as it does it for you based on past data.

Reviewer: I'm referring to that as well as the `freshness_cron` parameter for hardcoded freshness. When you use that parameter, instead of defining the amount of time between updates, you essentially define a deadline for when an update is expected by.

Author: Missed that, ty! Updated the copy here.

Using freshness checks, you can easily identify assets that are overdue for an update. As freshness checks don't depend on a specific root cause, it makes them helpful in accounting for unknowns. For example, freshness checks can identify overdue assets caused by any of the following:

- The pipeline hitting an error and failing
- Runs were never scheduled
- A backed up run queue
- Runs are taking longer than expected to complete

By the end of this guide, you'll understand what freshness checks are and how to implement them in your data pipelines.

---

## How it works

An asset is considered fresh if it's been materialized or observed within a defined time window. Otherwise, assets are considered overdue and in need of an update. To identify these assets, you can use **freshness checks**.

> Let's say an asset begins materializing at 12:00AM and takes 10 minutes to complete. Allowing for operational constraints and delays, the asset should always be materialized by 12:30AM. Using a freshness check, we can define the time that the asset should be updated. If it's not, Dagster will mark it as **Overdue**.

Reviewer: If the goal is provide a motivating example, I think a realistic one would be to talk about a situation where data is expected to be updated by 7 am every day, so that people reviewing dashboards will have access to up-to-date data.

Author: Good call!

How freshness checks are implemented depends on the type of asset they target:

> - **For source assets**, which are assets whose data feeds a Dagster pipeline, freshness checks utilize the time the asset was last updated to determine freshness. In this guide, we'll demonstrate how to use [observable source assets](/concepts/assets/asset-observations#observable-source-assets) and a [schedule](/concepts/partitions-schedules-sensors/schedules) to achieve this.

Reviewer: For both source assets and materializable assets, freshness checks utilize the time the asset was last updated. The difference is that the way "time the asset was last updated" is different for these two cases.

- **For materializable assets**, which are assets materialized by Dagster, Dagster can infer the asset's last update time using the asset's latest materialization timestamp. In this guide, we'll demonstrate how to use <PyObject object="asset" decorator />-decorated assets and a [sensor](/concepts/partitions-schedules-sensors/sensors) to achieve this.

For materializable assets, freshness checks can also help you communicate SLAs for their data freshness. For example, downstream consumers of the assets can look at the checks that are defined on those assets to determine when and how often they’re expected to be updated.

---

## Prerequisites

Before continuing, you should be familiar with:

- [Software-defined Assets](/concepts/assets/software-defined-assets)
- [Asset checks](/concepts/assets/asset-checks)
- [Schedules](/concepts/partitions-schedules-sensors/schedules) and/or [sensors](/concepts/partitions-schedules-sensors/sensors)

---

## Defining freshness checks for source assets

> In this section, we'll demonstrate how to implement freshness checks for source assets. Source assets are assets whose data feeds a Dagster pipeline, but isn't materialized by Dagster.

Reviewer: ```suggestion
In this section, we'll demonstrate how to implement freshness checks for source assets. Source assets are assets whose data feeds a Dagster pipeline, but aren't materialized by Dagster.
```

Author: 🤦‍♀️ I was definitely staring at this for way too long yesterday to have made this mistake. Thank you!

### Step 1: Track the asset's last update time

To run freshness checks on source assets, the checks need to know when the source assets were last updated. [Observable source assets](/concepts/assets/asset-observations#observable-source-assets) can be used to track the update times of these assets.

The following is an example of multiple Snowflake tables, backed by an observation function that queries Snowflake to find the most recent time the tables were updated. The function yields the update time as metadata to be stored in the Dagster event log:

```python

> from dagster import multi_observable_source_asset, AssetSpec, Definitions, MetadataValue

Reviewer: We should get these examples under test. After this gets merged, I or @dpeng817 should do a pass to extract them out into .py files.

Author: Agreed - I left a note about this to make sure it was accounted for. Additionally, I think I saw a note in the original Notion draft you wrote that a better example should be used for the materializable asset section - d'you think we can get that in and then do a fast-follow for moving everything?

from dagster_snowflake import fetch_last_updated_timestamps, SnowflakeResource

table_names = ["charges", "customers"]
asset_specs = [AssetSpec(table_name) for table_name in table_names]

@multi_observable_source_asset(specs=asset_specs)
def source_tables(snowflake: SnowflakeResource):
    with snowflake.get_connection() as conn:
        freshness_results = fetch_last_updated_timestamps(
            snowflake_connection=conn.cursor(), tables=table_names
        )
        for table_name, last_updated in freshness_results.items():
            yield ObserveResult(
                asset_key=table_name,
        metadata={"dagster/last_updated_timestamp": MetadataValue.timestamp(last_updated)},
            )
```

> ### Step 2: Create a schedule for the observation

Reviewer: ```suggestion
### Step 2: Schedule the observations
```

Next, we'll define a [schedule](/concepts/partitions-schedules-sensors/schedules) that regularly executes the function in the `source_tables` observable source asset:

```python
from dagster import ScheduleDefinition

source_tables_observation_schedule = ScheduleDefinition(
  job=define_asset_job(
    "source_tables_observation_job",
    selection=AssetSelection.assets(source_tables),
  ),
  # Runs every minute. Usually, a much less frequent cadence is necessary,
  # but a short cadence makes it easier to play around with this example.
  cron_schedule="* * * * *",
)
```

When the code location is loaded and the schedule is turned on, it will automatically kick off runs to observe the asset.

### Step 3: Define the freshness check

<TabGroup>
<TabItem name="Option 1: Use hardcoded parameters">

#### Option 1: Use hardcoded parameters

> In our example, we exepct the source tables to be updated no less than every two hours. We'll use the `build_freshness_checks` function to produce a set of asset checks that fail if an asset’s `last_updated_timestamp` is more than two hours before the current time:

Reviewer: ```suggestion
In our example, we expect the source tables to be updated no less than every two hours. We'll use the `build_freshness_checks` function to produce a set of asset checks that fail if an asset’s `last_updated_timestamp` is more than two hours before the current time:
```

Reviewer: `build_freshness_checks` should probably link to the API doc

```python
from dagster import build_freshness_checks_for_non_partitioned_assets

source_table_freshness_checks = build_freshness_checks_for_non_partitioned_assets(
  assets=[source_tables],
  maximum_lag_minutes=2 * 60,
)
```

These checks will automatically execute after the observations of the source assets they target, so an additional schedule isn't needed.

</TabItem>
<TabItem name="Option 2: Use time anomaly detection (Dagster Cloud Pro)">

> #### Option 2: Use time anomaly detection

Reviewer: ```suggestion
#### Option 2: Use anomaly detection (Dagster Cloud Pro-only)
```

<Note>A Dagster Cloud Pro plan is required to use this feature.</Note>

TODO: ADD NOTE ABOUT MATERIALIZATIONS AND ACCURACY

Setting custom freshness policies on a large number of source assets can be time-consuming. Dagster Cloud Pro users can take advantage of a time series anomaly detection model instead of applying policies on an asset-by-asset basis. Freshness checks that use this approach function the same way checks with hardcoded parameters do.

This model uses data from past materializations/observations to determine if data is arriving later than expected. **Note**: If the asset hasn't been updated enough times, the check will pass with a message indicating that more data is needed to detect anomalies.

> In the following example, we used `build_anomaly_detection_freshness_checks` to accomplish this:

Reviewer: ```suggestion
In the following example, we use `build_anomaly_detection_freshness_checks` to accomplish this:
```

```python
from dagster_cloud import build_anomaly_detection_freshness_checks

freshness_checks = build_anomaly_detection_freshness_checks(
  assets=[source_tables], params=None
)
```

</TabItem>
</TabGroup>

> ### Step 4: Update the Definitions object

Reviewer: ```suggestion
### Step 4: Create a Definitions object
```

Reviewer: ^^suggested because there's not a preexisting Definitions object that they reupdating.

        )
        for table_name, last_updated in freshness_results.items():
            yield ObserveResult(
                asset_key=table_name,
        metadata={"dagster/last_updated_timestamp": MetadataValue.timestamp(last_updated)},
            )

source_tables_observation_schedule = ScheduleDefinition(
  job=define_asset_job(
    "source_tables_observation_job",
    selection=AssetSelection.assets(source_tables),
  ),
  cron_schedule="* * * * *",
)

source_table_freshness_checks = build_freshness_checks_for_non_partitioned_assets(
  assets=[source_tables],
  maximum_lag_minutes=2 * 60,
)

defs = Definitions(
  assets=[source_table_freshness_checks],
  asset_checks=[source_table_freshness_checks],
  schedules=[source_tables_observation_schedule],
)
```

From here, you can [view and manually execute the checks in the Dagster UI](#freshness-checks-in-the-dagster-ui).

---

## Defining freshness checks for materializable assets

In this section, we'll demonstrate how to implement freshness checks for materializable assets. These are assets that are materialized by a Dagster pipeline.

### Step 1: Define the freshness check

> In this example, we used the `build_freshness_checks` function to produce an asset check that will fail if an asset’s latest materialization is more than two hours before the current time:

Reviewer: ```suggestion
In this example, we use the `build_freshness_checks` function to produce an asset check that will fail if an asset’s latest materialization is more than two hours before the current time:
```

```python
from dagster import build_freshness_checks_for_non_partitioned_assets

@asset
def my_asset():
  ...

asset1_freshness_checks = build_freshness_checks_for_non_partitioned_assets(
  assets=[my_asset], maximum_lag_minutes=2 * 60
)
```

> ### Step 2: Create a sensor to excecute the check

Reviewer: ```suggestion
### Step 2: Create a sensor to execute the check
```

---
title: "Using asset checks to check data freshness | Dagster Docs"
description: "Use freshness checks to identify data assets that are overdue for an update."
---

# Using asset checks to check data freshness

A freshness check is a type of [asset check](/concepts/assets/asset-checks) that allows you identify Dagster [assets](/concepts/assets/software-defined-assets) that are overdue for a data refresh.

As freshness checks don't depend on a specific root cause, it makes them helpful in accounting for unknowns. For example, freshness checks can identify stale assets caused by any of the following:

- The pipeline hitting an error and failing
- Runs were never scheduled
- A backed up run queue
- Runs are taking longer than expected to complete

By the end of this guide, you'll understand what freshness checks are and how to implement them in your data pipelines.

---

## How it works

An asset is considered fresh if it's been materialized or observed within a defined time window. Otherwise, assets are considered overdue and in need of an update. To identify these assets, you can use **freshness checks**.

> Let's say members of your team review a dashboard every morning as part of their daily stand up. To ensure they have the most up-to-date information, assets powering the dashboard should be materialized no later than 7AM. Using a freshness check, we can define the time that the asset should be updated. If it's not, Dagster will mark it as **Overdue**.

Reviewer: ```suggestion
Let's say members of your team review a dashboard every morning as part of their daily stand up. To ensure they have the most up-to-date information, assets powering the dashboard should be materialized no later than 7AM. Using a freshness check, we can define the time that the asset should be updated by. If the check runs at that time and the asset hasn't been updated, the asset will be considered overdue and the check will fail.
```

Additionally, freshness checks can also help communicate SLAs for their data freshness. For example, downstream asset consumers can look at the checks that are defined on those assets to determine when and how often they’re expected to be updated.

Freshness checks use an asset's latest update time to determine if it's overdue. How Dagster determines the last update time varies on the type of asset being targeted:

- **For source assets**, which are assets whose data feeds a Dagster pipeline, freshness checks utilize the time the asset was last updated to determine freshness. In this guide, we'll demonstrate how to use [observable source assets](/concepts/assets/asset-observations#observable-source-assets) and a [schedule](/concepts/partitions-schedules-sensors/schedules) to achieve this.

> - **For materializable assets**, which are assets materialized by Dagster, Dagster infers the asset's last update time using the its latest materialization timestamp. In this guide, we'll demonstrate how to use <PyObject object="asset" decorator />-decorated assets and a [sensor](/concepts/partitions-schedules-sensors/sensors) to achieve this.

Reviewer: Maybe worth linking to the relevant section below?

        )
        for table_name, last_updated in freshness_results.items():
            yield ObserveResult(
                asset_key=table_name,
        metadata={"dagster/last_updated_timestamp": MetadataValue.timestamp(last_updated)},
            )
```

### Step 2: Schedule the observations

Next, we'll define a [schedule](/concepts/partitions-schedules-sensors/schedules) that regularly executes the function in the `source_tables` observable source asset:

```python
from dagster import define_asset_job, AssetSelection, ScheduleDefinition

source_tables_observation_schedule = ScheduleDefinition(
  job=define_asset_job(
    "source_tables_observation_job",
    selection=AssetSelection.assets(source_tables),
  ),
  # Runs every minute. Usually, a much less frequent cadence is necessary,
  # but a short cadence makes it easier to play around with this example.
  cron_schedule="* * * * *",
)
```

When the code location is loaded and the schedule is turned on, it will automatically kick off runs to observe the asset.

### Step 3: Define the freshness check

<TabGroup>
<TabItem name="Option 1: Use hardcoded parameters">

#### Option 1: Use hardcoded parameters

> In our example, we exepct the source tables to be updated no less than every two hours. We'll use the <PyObject object="build_freshness_checks_for_non_partitioned_assets" /> function to produce a set of asset checks that fail if an asset’s `last_updated_timestamp` is more than two hours before the current time:

Reviewer: ```suggestion
In our example, we expect the source tables to be updated no less than every two hours. We'll use the <PyObject object="build_freshness_checks_for_non_partitioned_assets" /> function to produce a set of asset checks that fail if an asset’s `last_updated_timestamp` is more than two hours before the current time:
```

```python
from dagster import build_freshness_checks_for_non_partitioned_assets

source_table_freshness_checks = build_freshness_checks_for_non_partitioned_assets(
  assets=[source_tables],
  maximum_lag_minutes=2 * 60,
)
```

These checks will automatically execute after the observations of the source assets they target, so an additional schedule isn't needed.

</TabItem>
<TabItem name="Option 2: Use time anomaly detection (Dagster Cloud Pro)">

> #### Option 2: Use time anomaly detection (Dagster Cloud Pro)

Reviewer: ```suggestion
#### Option 2: Use anomaly detection (Dagster Cloud Pro)
```
